File: problems_list/0300.最长递增子序列/0300-最长递增子序列.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n=nums.size(), len=1;
        if(n==0) return 0;

        vector<int> d(n+1, 0);
        d[len]=nums[0];
        for(int i=1; i<n; i++){
            if(nums[i]>d[len]){
                ++len;
                d[len]=nums[i];
            }else{
                int l=1, r=len, pos=0;
                while(l<=r){
                    int m=(r+l)>>1;
                    if(d[m]<nums[i]){
                        pos=m;
                        l=m+1;
                    }else{
                        r=m-1;
                    }
                }
                d[pos+1]=nums[i];
            }
        }
        return len;
    }
};

File: problems_list/0300.最长递增子序列/0300-最长递增子序列.js
/**
 * @param {number[]} nums
 * @return {number}
 */

var lengthOfLIS = function(nums) {
    let n=nums.length, len=1;
    if(n===0) return 0;

    let dp=new Array(n+1).fill(0);
    dp[len]=nums[0];
    for(let i=1; i<n; i++){
        if(nums[i]>dp[len]){
            len++;
            dp[len]=nums[i];
        }else{
            // 第一个大于等于 nums[i]
            let l=1, r=len, pos=0;
            while(l<=r){
                let m=Math.floor((l+r)>>1);
                if(dp[m]<nums[i]){
                    pos=m;
                    l=m+1;
                }else{
                    r=m-1;
                }
            }
            dp[pos+1]=nums[i];
        }
    }
    return len;
};